# TD 1 | Introduction à Python pour l'analyse de données

---

Objectifs du TD :

* découvrir Python
* se familiariser avec le langage et le notebook Jupyter
* découvrir et maîtriser les bases des librairies de calcul numérique et d'analyse de données numpy et pandas

---

##  Installation de Python et Jupyter

### Windows

#### Installation Python

Vérifiez si vous avez déjà une version de python installée en ouvrant l'invite de commande et en exécutant l'une des deux commandes python suivantes : 
> python --version   

> python3 --version 

Si ce n'est pas le cas, vous pouvez procéder à l'installation comme suit :
 
1. Allez sur le site web : https://www.python.org/downloads/windows/
2. Cliquez sur "Latest Python 3 Release - Python 3.10.7".
3. Si votre ordinateur utilise une version 64 bits de Windows, téléchargez le Windows installer (64-bit). Sinon, téléchargez Windows installer (32-bit).
4. Après avoir téléchargé le Windows installer, vous devez l'exécuter (double-cliquez dessus) et suivre les instructions qui s'y trouvent.

NB : Au début de l'installation, assurez-vous de cocher la case "Add Python 3.10 to PATH" ou "Add Python to your environment variables" avant de cliquer sur "Install Now".


#### Installation Jupyter

Ouvrir l'invite de commande et exécuter la commande suivante en remplaçant Pseudo_de_votre_user par le nom de l'utilisateur de votre session : 
> cd "C:\Users\Pseudo_de_votre_user\AppData\Local\Programs\Python\Python310\Scripts"

Ensuite, executer la commande suivante :
> pip install jupyterlab

Une fois l'installation terminée avec succès, vous pouvez lancer JupyterLab avec la commande suivante : 
> jupyter-lab


### Linux & Mac OS

#### Installation Python

Normalement, Python est préinstallé sur la quasi-totalité des distributions. Vous n'avez pas besoin de le réinstaller, mais vous pouvez mettre à jour la version si vous le souhaitez.

Vérifiez la version Python installée en exécutant l'une des deux commandes python suivantes sur votre terminal : 
> python --version   

> python3 --version 

#### Installation Jupyter

Vérifier que la commande pip3 est installée. Sinon, l'installer (exemple pour Debian/Ubuntu/Linux Mint, à adapter selon votre distribution) :

> sudo apt install python3-pip

Ensuite, installer JupyterLab en executant la commande suivante:
> pip3 install -U jupyterlab

Enfin, placez-vous dans votre dossier de travail et lancez un notebook :

> jupyter-lab

## Syntaxe et Concepts de base

In [269]:
a = 33
b = 9
c = 2.0

In [270]:
a+b

42

In [271]:
print(a+b)

42


In [272]:
type(a)

int

In [273]:
type(c)

float

In [274]:
a = 'hello'
type(a)

str

In [275]:
# Affiche les entiers pairs de 1 à 10
for i in range(1, 11):
    if i % 2 == 0:
        print(i)

2
4
6
8
10


### List

In [276]:

l = [1,2,3,4,4]

#### len, max, min

In [277]:
print('longueur de la liste =', len(l))
print("Valeur max de l =", max(l))
print("Valeur min de l =", min(l))

longueur de la liste = 5
Valeur max de l = 4
Valeur min de l = 1


#### Ajout d'un élément d'une liste

In [278]:


l.append(6)
l

[1, 2, 3, 4, 4, 6]

In [279]:
l = l+[7]
l

[1, 2, 3, 4, 4, 6, 7]

In [280]:
l+=[7]
l

[1, 2, 3, 4, 4, 6, 7, 7]


#### Modifier un élément

In [281]:
l[0] = 29
l

[29, 2, 3, 4, 4, 6, 7, 7]

In [282]:

l.insert(1,9)
l

[29, 9, 2, 3, 4, 4, 6, 7, 7]

#### Suppression des éléments d'une liste

In [283]:
l.remove(7)
l

[29, 9, 2, 3, 4, 4, 6, 7]

In [284]:



l.pop(2)
l

[29, 9, 3, 4, 4, 6, 7]

#### Trier une liste

In [285]:
l.sort(reverse=True)
print(l)

[29, 9, 7, 6, 4, 4, 3]


In [286]:

l_sorted = sorted(l)
print('Sorted list:', l_sorted)

Sorted list: [3, 4, 4, 6, 7, 9, 29]




#### Liste non homogène

In [287]:
l = [1, 'un', 2]
l.append('deux')
l += [3, 'trois']
print(type(l))
l

<class 'list'>


[1, 'un', 2, 'deux', 3, 'trois']

In [288]:
a = l[3], l[-1]
print(type(a))
a

<class 'tuple'>


('deux', 'trois')

In [289]:
a[0]

'deux'

In [290]:
[2**p for p in range(10)] # "Compréhension de liste"

[1, 2, 4, 8, 16, 32, 64, 128, 256, 512]

In [291]:
# Liste des entiers pairs de 1 à 10
l = [i for i in range(1, 11) if i % 2 == 0]
print(l)

[2, 4, 6, 8, 10]


### Tuple

In [292]:
p = tuple(range(1,5))
print("p=", p)
print("p[0] = ", p[0])

p= (1, 2, 3, 4)
p[0] =  1


In [293]:
p = ([1,2],5)

print(p)

p[0][0] = 8

print(p)

([1, 2], 5)
([8, 2], 5)


In [294]:
# Les éléments du tuple sont fixes
p[0] = 6

TypeError: 'tuple' object does not support item assignment

In [ ]:
p1 = (1,2)
p2 = (2,1)

print(p1 == p2)

False


### Set

In [ ]:
s = set((1,2,'l'))
print(s)

s = {4,5,9,9,9,9}
print(s)

{1, 2, 'l'}
{9, 4, 5}


In [ ]:
#Ajout d'un élément au set
s.add(6)
print(s)

{9, 4, 5, 6}


In [ ]:
#Suppression d'un élément du set
s.discard(5)
print(s)

{9, 4, 6}


In [ ]:
s.clear()
print(s)

set()


In [ ]:
A = {1,2,3,4,5}
B = {2,4,6,7}
C = {1,2,3}

In [ ]:
#Intersection
print("Intersection de A et B =", A&B)

#Union
print("Union entre A et B =", A|B)

#Exclusion
print("Exclusion de B de A =", A-B)

#Différence symétrique
print("Diff sym. A et B =", A^B)

Intersection de A et B = {2, 4}
Union entre A et B = {1, 2, 3, 4, 5, 6, 7}
Exclusion de B de A = {1, 3, 5}
Diff sym. A et B = {1, 3, 5, 6, 7}


In [ ]:
print("B est inclus strict dans A") if A>B else print(False)
print("C est inclus dans A") if A>=C else print(False)

False
C est inclus dans A


### Dictionary

In [ ]:
# 1ère façon de créer un dictionnaire
dic = {"key1": "value1", "answer": 42}
# 2ème façon de créer un dictionnaire
dic2 = dict(key1="value1", answer=42)
# 3ème façon de créer un dictionnaire
dic3 = dict([("key1","value1"), ("answer",42)])
print(dic)
print(dic2)
print(dic3)

dic["answer"]

{'key1': 'value1', 'answer': 42}
{'key1': 'value1', 'answer': 42}
{'key1': 'value1', 'answer': 42}


42

In [ ]:
dic['new'] = [1,2,3]

In [ ]:
dic.keys()

dict_keys(['key1', 'answer', 'new'])

In [ ]:
dic.items()

dict_items([('key1', 'value1'), ('answer', 42), ('new', [1, 2, 3])])

In [ ]:
dic.values()

dict_values(['value1', 42, [1, 2, 3]])

In [ ]:
a,b,c = dic.items()
print(a)

('key1', 'value1')


## Programmation Orientée Objet

In [ ]:
"""
Exemple de classe en Python
"""
class Moteur:
    
    # Constructeur
    def __init__(self, esn, panne=False):
        self.esn = esn
        self.panne = panne
    
    # Méthodes
    def dire_bonjour(self):
        print('Bonjour, mon numéro de série est ' + self.esn)
    
    def fonctionne(self):
        return not self.panne 
    

In [ ]:
mot1 = Moteur('420912')
mot2 = Moteur(panne=True, esn='420913')

mot1.dire_bonjour()
print(mot1.fonctionne())

mot1.panne = True
print(mot1.fonctionne())

print('\n')
mot2.dire_bonjour()
print(mot2.fonctionne())

Bonjour, mon numéro de série est 420912
True
False


Bonjour, mon numéro de série est 420913
False


### Méthode récurcive vs Méthode itérative

In [ ]:
# Méthode récursive
def fact1(n):
    if n == 0:
        return 1
    else:
        return n*fact1(n-1)

# Méthode itérative
def fact2(n):
    s=1
    for i in range(1,n+1):
        s*=i
    return s

print(fact1(5))
print(fact2(5))

120
120


In [ ]:
"""
App.1 : Ecrire une fonction qui retourne le plus grand commun diviseur (pgcd) entre deux nombres a et b.

"""

def pgcd(a,b):
    M=max(a,b)
    N=min(a,b)
    c=M-N
    while c !=0:
        M,N=max(c,N),min(c,N)
        c=M-N
    res=M    
    return res
    

In [ ]:
pgcd(230,25)

5

In [ ]:
"""
App.1 : Correction
"""

#Méthode itérative
def pgcd1(a,b):
    while (a%b) != 0:
        s = a%b
        a = b
        b = s
    return b


#Méthode récurcive
def pgcd2(a,b):
    if a==b : return a 
    if a<b : a,b = b,a
    return pgcd2(a-b,b)

a = 230
b = 25

print("pgcd1(a,b) = ", pgcd1(a,b))
print("pgcd2(a,b) = ", pgcd2(a,b))

pgcd1(a,b) =  5
pgcd2(a,b) =  5


## Libraries

> pip install numpy pandas

> pip3 install -U numpy pandas

### Math

In [ ]:
from math import *

In [ ]:
x = 4

print("sqrt(x) =", sqrt(x))
print("cos(x) =", cos(x))
print("sin(x) =", sin(x))
print("factorial(x) =", factorial(x))

sqrt(x) = 2.0
cos(x) = -0.6536436208636119
sin(x) = -0.7568024953079283
factorial(x) = 24


### Numpy

In [ ]:
import numpy as np

In [ ]:
# Création d'un array à partir d'une liste
v = np.array([1.0, 2.0, 3.0])
# Création d'un array de taille (n,m) initialisé à 0
z = np.zeros((3,4))
z

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [ ]:
# Taille d'un array
print(z.shape)
print(v.shape)

(3, 4)
(3,)


In [ ]:
# Opérations courantes
A = np.ones((3,3)) + np.eye(3)
print('A = ', A)
print('Av = ', np.dot(A,v))
print('3*A = ', 3*A)
print('A*v = ', A*v)
print('A + 1 = ', A+1)
print('A + v = ', A+v)
print('A^2 = ', np.square(A))

A =  [[2. 1. 1.]
 [1. 2. 1.]
 [1. 1. 2.]]
Av =  [7. 8. 9.]
3*A =  [[6. 3. 3.]
 [3. 6. 3.]
 [3. 3. 6.]]
A*v =  [[2. 2. 3.]
 [1. 4. 3.]
 [1. 2. 6.]]
A + 1 =  [[3. 2. 2.]
 [2. 3. 2.]
 [2. 2. 3.]]
A + v =  [[3. 3. 4.]
 [2. 4. 4.]
 [2. 3. 5.]]
A^2 =  [[4. 1. 1.]
 [1. 4. 1.]
 [1. 1. 4.]]


In [ ]:
B = np.array([5, 2, 9, 1])

print(np.sort(B))

[1 2 5 9]


#### Compatif de performance : produit matriciel


In [ ]:
import random

In [ ]:
# Création d'une matrice aléatoire sous forme de liste de listes
taille = 2
A = [[random.random() for _ in range(taille)] for _ in range(taille)]

In [ ]:
print(A)

[[0.8799401820367159, 0.10815766259186887], [0.4011125024611588, 0.2705347273098234]]


In [ ]:
"""
EXERCICE - Afficher un tuple contenant les dimensions de la matrice A
"""
print(np.shape(A))

(2, 2)


$$ (AB)_{ij} = \sum_k A_{ik} B_{kj} $$

In [ ]:
"""
EXERCICE - Implémenter le produit matriciel de 2 matrices sous forme de listes de listes python
"""
def produit(A,B):
    n1,p1,n2,p2=np.shape(A)[0], np.shape(A)[1] ,np.shape(B)[0], np.shape(B)[1]
    res=np.zeros((n1,p2))
    if p1 != n2  :
        raise NotImplementedError
    else:
        for i in range(n1):
            for j in range(p2):
                for k in range(p1):
                    res[i][j]+= A[i][k]*B[k][j]
    return res.tolist()
                        
            
         
    
        
    


In [ ]:
list(np.zeros((4,5)))

[array([0., 0., 0., 0., 0.]),
 array([0., 0., 0., 0., 0.]),
 array([0., 0., 0., 0., 0.]),
 array([0., 0., 0., 0., 0.])]

In [ ]:
# Vérification
assert(produit([[1, 2], [3, 4]], [[1, 2], [3, 4]]) == [[7, 10], [15, 22]])

In [ ]:
produit(A, A)

[[0.8176781146653829, 0.12443267706551393],
 [0.4614698699067652, 0.1165724293831747]]

In [ ]:
# Sortons le chronomètre
%timeit produit(A, A)

5.23 µs ± 52.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [ ]:
# Et maintenant, avec numpy !
import numpy as np

In [ ]:
A2 = np.array(A)
print(A2)

[[0.87994018 0.10815766]
 [0.4011125  0.27053473]]


In [ ]:
# Création d'une matrice aléatoire sous forme de liste de listes
taille = 30
A = [[random.random() for _ in range(taille)] for _ in range(taille)]
# Sortons le chronomètre
%timeit produit(A, A)

4.79 ms ± 63.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [ ]:
%timeit np.dot(A, A)

63.8 µs ± 670 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


EXERCICE - Remplissez le tableau suivant avec les durées d'exécution constatées :

(Conseil : pour la taille 3000, essayez UNIQUEMENT avec numpy)

Taille | Python             | numpy
-------|--------            |-------
30     |4.65 ms ± 16.8 µs         | 63.8µs ± 945ns
300    |4.88 s ± 91.8 ms         | 21.3ms ±1.24ms 
3000   |         | 1.09 s ±61.3 ms

###  Pandas

In [ ]:
import pandas as pd

In [ ]:
df_exemple = pd.DataFrame({"ESN": ["E420912", "E420913", "E420914"], "panne": [False, False, True]})
df_exemple

ESN  panne
0  E420912  False
1  E420913  False
2  E420914   True

#### Lecture et prétraitement de données


In [ ]:
!git clone https://github.com/MadaneA/MACS3-Statistiques-Descriptives-TDs.git

fatal: destination path 'MACS3-Statistiques-Descriptives-TDs' already exists and is not an empty directory.


In [ ]:
import os
os.chdir('MACS3-Statistiques-Descriptives-TDs')

In [ ]:
 os.getcwd()

'/Users/renaud/Desktop/Laptop_Ubuntu/MACS3/Statistiques/TD/MACS3-Statistiques-Descriptives-TDs-main/TD1/MACS3-Statistiques-Descriptives-TDs'

In [ ]:
# Chargement d'un fichier CSV ou Excel
df = pd.read_csv("./data/Vol010.csv")
# Affichage des 5 premières lignes
df.head()

t EGT_SEL FLIGHT_MOD FMV_SEL HPTC_SEL LPTC_SEL  \
0                  datenum   deg_C          _       _        %        %   
1  15/09/2011 14:25:58.125     NaN        NaN     NaN      NaN      NaN   
2  15/09/2011 14:25:58.375     NaN        NaN     NaN      NaN      NaN   
3  15/09/2011 14:25:58.625     NaN        NaN     NaN      NaN      NaN   
4  15/09/2011 14:25:58.875     NaN        NaN     NaN      NaN      NaN   

  N1_SEL N2_ACTSEL OIL_P OIL_TEMP  ... T25_SEL T3_SEL VBV_SEL VIB_CN1 VIB_CN2  \
0      %     %_RPM   psi        _  ...       _      _     DEG       _       _   
1    NaN         0   NaN      NaN  ...     NaN    NaN     NaN     NaN     NaN   
2    NaN         0   NaN      NaN  ...     NaN    NaN     NaN     NaN     NaN   
3    NaN         0   NaN      NaN  ...     NaN    NaN     NaN     NaN     NaN   
4    NaN         0   NaN      NaN  ...     NaN    NaN     NaN     NaN     NaN   

  VIB_TN1 VIB_TN2 VSV_SEL  WFM_SEL    XM  
0       _       _     DEG     lb/h  mach  
1     NaN     NaN     NaN      NaN   NaN  
2     NaN     NaN     NaN      NaN   NaN  
3     NaN     NaN     NaN      NaN   NaN  
4     NaN     NaN     NaN  7679.98  0.15  

[5 rows x 26 columns]

In [ ]:
"""
EXERCICE - Dimensions d'un DataFrame
Affichez le nombre de colonnes et de ligne du DataFrame (indice : beaucoup de méthodes sont communes entre numpy et pandas)
"""
print('Nombre de colonnes :', df.shape[1] )
print('Nombre de lignes :',df.shape[0] )

Nombre de colonnes : 26
Nombre de lignes : 22945


In [ ]:
df.size, len(df)

(596570, 22945)

In [ ]:
df.columns

Index(['t', 'EGT_SEL', 'FLIGHT_MOD', 'FMV_SEL', 'HPTC_SEL', 'LPTC_SEL',
       'N1_SEL', 'N2_ACTSEL', 'OIL_P', 'OIL_TEMP', 'PS3_SEL', 'PT2_SEL',
       'P0_SEL', 'TAT', 'TBV_SEL', 'TRA_SEL', 'T25_SEL', 'T3_SEL', 'VBV_SEL',
       'VIB_CN1', 'VIB_CN2', 'VIB_TN1', 'VIB_TN2', 'VSV_SEL', 'WFM_SEL', 'XM'],
      dtype='object')

Les colonnes des DF sont typées, à la manière d'une base de données relationnelle, contrairement aux variables python classiques. Les types des colonnes sont accessibles via df.dtypes. Les principaux types sont les numériques (int32, int64, float etc.

In [ ]:
df.loc[10:15, ['EGT_SEL', 'FLIGHT_MOD']]

EGT_SEL FLIGHT_MOD
10    0.00234327        NaN
11  1.08032e-007        NaN
12  3.12514e-012        NaN
13             0        NaN
14       72.0006        NaN
15       215.996        NaN

In [ ]:
"""
EXERCICE - Extraction et suppression des unités
On remarque que la 1ère ligne ne contient pas de données mais les unités de chaque colonne.
Pour la suite des traitements, il faut supprimer cette ligne. On souhaite toutefois garder l'information des unités de chaque colonne.
1. Récupérez les unités et stockez les dans une structure adaptée.
2. Supprimez cette ligne du DataFrame en utilisant la méthode "drop"
"""
units=list(df.loc[0,:]) 
units
df=df.drop(0)

In [ ]:
help(df.drop)

Help on method drop in module pandas.core.frame:

drop(labels: 'IndexLabel' = None, *, axis: 'Axis' = 0, index: 'IndexLabel' = None, columns: 'IndexLabel' = None, level: 'Level' = None, inplace: 'bool' = False, errors: 'IgnoreRaise' = 'raise') -> 'DataFrame | None' method of pandas.core.frame.DataFrame instance
    Drop specified labels from rows or columns.
    
    Remove rows or columns by specifying label names and corresponding
    axis, or by specifying directly index or column names. When using a
    multi-index, labels on different levels can be removed by specifying
    the level. See the :ref:`user guide <advanced.shown_levels>`
    for more information about the now unused levels.
    
    Parameters
    ----------
    labels : single label or list-like
        Index or column labels to drop. A tuple will be used as a single
        label and not treated as a list-like.
    axis : {0 or 'index', 1 or 'columns'}, default 0
        Whether to drop labels from the index (0 or '

In [ ]:
print(units)
df.head()

['datenum', 'deg_C', '_', '_', '%', '%', '%', '%_RPM', 'psi', '_', 'psi', 'mb', 'psi', 'deg_C', '%', 'DEG', '_', '_', 'DEG', '_', '_', '_', '_', 'DEG', 'lb/h', 'mach']


t EGT_SEL FLIGHT_MOD FMV_SEL HPTC_SEL LPTC_SEL  \
1  15/09/2011 14:25:58.125     NaN        NaN     NaN      NaN      NaN   
2  15/09/2011 14:25:58.375     NaN        NaN     NaN      NaN      NaN   
3  15/09/2011 14:25:58.625     NaN        NaN     NaN      NaN      NaN   
4  15/09/2011 14:25:58.875     NaN        NaN     NaN      NaN      NaN   
5  15/09/2011 14:25:59.125     NaN        NaN     NaN      NaN      NaN   

  N1_SEL N2_ACTSEL    OIL_P OIL_TEMP  ... T25_SEL T3_SEL VBV_SEL VIB_CN1  \
1    NaN         0      NaN      NaN  ...     NaN    NaN     NaN     NaN   
2    NaN         0      NaN      NaN  ...     NaN    NaN     NaN     NaN   
3    NaN         0      NaN      NaN  ...     NaN    NaN     NaN     NaN   
4    NaN         0      NaN      NaN  ...     NaN    NaN     NaN     NaN   
5      0         0  647.998      NaN  ...     NaN    NaN       0       0   

  VIB_CN2 VIB_TN1 VIB_TN2 VSV_SEL  WFM_SEL    XM  
1     NaN     NaN     NaN     NaN      NaN   NaN  
2     NaN     NaN     NaN     NaN      NaN   NaN  
3     NaN     NaN     NaN     NaN      NaN   NaN  
4     NaN     NaN     NaN     NaN  7679.98  0.15  
5       0       0       0       0  4607.86  0.15  

[5 rows x 26 columns]

On remarque que toutes les colonnes ont été reconnues comme de type object, c'est-à-dire des chaînes de caractères, alors que ce sont des valeurs numériques. Cela est dû à la première ligne contenant les unités. Il faut donc convertir les colonnes en numérique. La colonne 't', quant à elle, doit être convertie en type datetime.

In [ ]:
df['t'] = pd.to_datetime(df['t'])
df[df.columns[1:]] = df[df.columns[1:]].apply(pd.to_numeric)

/var/folders/gh/8sgjg9tj54v1mgxh73dkqv_h0000gn/T/ipykernel_97131/2306591669.py:1: UserWarning: Parsing dates in %d/%m/%Y %H:%M:%S.%f format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df['t'] = pd.to_datetime(df['t'])


In [ ]:
df.dtypes

t             datetime64[ns]
EGT_SEL              float64
FLIGHT_MOD           float64
FMV_SEL              float64
HPTC_SEL             float64
LPTC_SEL             float64
N1_SEL               float64
N2_ACTSEL            float64
OIL_P                float64
OIL_TEMP             float64
PS3_SEL              float64
PT2_SEL              float64
P0_SEL               float64
TAT                  float64
TBV_SEL              float64
TRA_SEL              float64
T25_SEL              float64
T3_SEL               float64
VBV_SEL              float64
VIB_CN1              float64
VIB_CN2              float64
VIB_TN1              float64
VIB_TN2              float64
VSV_SEL              float64
WFM_SEL              float64
XM                   float64
dtype: object

In [ ]:
"""
Exercice - Index temporel
Comme nos données sont une série temporelle multivariée, on souhaite utiliser un index temporel.
1. Créez une copie de df, appelée df2, à l'aide de la méthode du même nom.
2. Affectez la colonne du temps ('t') en tant qu'indice du DataFrame.
3. Supprimez la colonne 't' du DF résultant.
"""
# copie de df dans df2
df2=df.copy()

# affectation de la colonne du temps en tant qu'indice de df2

df2.index=df['t'] 

#suppression de la colonne "t"
df2=df2.drop(["t"], axis=1)



On constate que pandas a automatiquement reconnu un DatetimeIndex, adapté pour des manipulations de séries temporelles (moyennes glissantes, etc) !

### Valeurs manquantes

**NaN = Not a Number**

Les valeurs NaN doivent être éliminées ou imputées (i.e. remplacées par une certaine valeur) avant la suite des traitements. Ce choix dépend du cas d'usage. Dans un premier temps, nous allons apprendre à :

* trouver les données manquantes (méthode `isna`)
* éliminer les données manquantes d'un DataFrame (méthode `dropna`)
* les remplacer par une constante (méthode (`fillna`)

In [ ]:
help(df.any)

Help on method any in module pandas.core.generic:

any(*, axis: 'Axis' = 0, bool_only=None, skipna: 'bool_t' = True, **kwargs) method of pandas.core.frame.DataFrame instance
    Return whether any element is True, potentially over an axis.
    
    Returns False unless there is at least one element within a series or
    along a Dataframe axis that is True or equivalent (e.g. non-zero or
    non-empty).
    
    Parameters
    ----------
    axis : {0 or 'index', 1 or 'columns', None}, default 0
        Indicate which axis or axes should be reduced. For `Series` this parameter
        is unused and defaults to 0.
    
        * 0 / 'index' : reduce the index, return a Series whose index is the
          original column labels.
        * 1 / 'columns' : reduce the columns, return a Series whose index is the
          original index.
        * None : reduce all axes, return a scalar.
    
    bool_only : bool, default None
        Include only boolean columns. If None, will attempt to us

In [ ]:
"""
EXERCICE - La méthode isna
1. Testez la méthode isna sur le DataFrame df2, puis sur une colonne ou une ligne. Que renvoie-t-elle ?
2. En appliquant les méthodes any(axis=...), mean() et max()/idxmax() sur les résultats de isna(), répondez aux questions suivantes :
    2.1 Quelles colonnes contiennent des valeurs manquantes, lesquelles n'en contiennent pas ?
    2.2 Quel est le pourcentage de valeurs manquantes dans le DF (a) par colonne (b) globalement ? Quelle variable contient le plus de NaN ?
    2.3 Quel est le pourcentage d'indices du DF pour lesquels toutes les variables sont présentes ?
"""

#Résultat du test de la méthode isna sur df2: 
# La méthode isna renvoi une dataframe ou une serie de  booléen ("True" dans les cellules contenant NaN et "false" dans les autres. )



print('Question 2.1')
colNaN=df2.isna().any(axis=0) #Renvoi une serie de booléens  associés chaqu'un à chaque colonne ("True" si la colonne contient une valeur manquante).
print(f" \nles colonnes contenant au moins une valeur manquante: \n",colNaN[colNaN==True].index) #Renvoi les noms des colones contenant une valeur manquante.
print(f" \nles colonnes ne contenant aucune valeur manquante: \n",colNaN[colNaN==False].index) #Renvoi les noms des colones contenant une valeur manquante.
print('Question 2.2')
print("le pourcentage de valeurs manquantes dans le DF  par colonne est : \n")
print((df2.isna()*1).mean(axis=0))
print(f"\n le pourcentage de valeurs manquantes dans le DF  globalement est : \n", (df2.isna()*1).mean(axis=0).sum())
print(f"\n La variable contenant le plus de NaN  est :\n ",(df2.isna()*1).mean(axis=0).idxmax())
print('Question 2.3')
print(f" \n le pourcentage d'indices du DF pour lesquels toutes les variables sont présentes est : \n", ((df2.isna()*1).mean(axis=1)==1).mean())

Question 2.1
 
les colonnes contenant au moins une valeur manquante: 
 Index(['EGT_SEL', 'FLIGHT_MOD', 'FMV_SEL', 'HPTC_SEL', 'LPTC_SEL', 'N1_SEL',
       'N2_ACTSEL', 'OIL_P', 'OIL_TEMP', 'PS3_SEL', 'PT2_SEL', 'P0_SEL', 'TAT',
       'TBV_SEL', 'TRA_SEL', 'T25_SEL', 'T3_SEL', 'VBV_SEL', 'VIB_CN1',
       'VIB_CN2', 'VIB_TN1', 'VIB_TN2', 'VSV_SEL', 'WFM_SEL', 'XM'],
      dtype='object')
 
les colonnes ne contenant aucune valeur manquante: 
 Index([], dtype='object')
Question 2.2
le pourcentage de valeurs manquantes dans le DF  par colonne est : 

EGT_SEL       0.003487
FLIGHT_MOD    0.004446
FMV_SEL       0.004446
HPTC_SEL      0.004228
LPTC_SEL      0.004576
N1_SEL        0.004184
N2_ACTSEL     0.002528
OIL_P         0.003792
OIL_TEMP      0.003835
PS3_SEL       0.003007
PT2_SEL       0.003705
P0_SEL        0.003705
TAT           0.003007
TBV_SEL       0.002964
TRA_SEL       0.003007
T25_SEL       0.004707
T3_SEL        0.004707
VBV_SEL       0.002920
VIB_CN1       0.002920
VIB_CN2  

In [ ]:
help(df.dropna)

Help on method dropna in module pandas.core.frame:

dropna(*, axis: 'Axis' = 0, how: 'AnyAll | NoDefault' = <no_default>, thresh: 'int | NoDefault' = <no_default>, subset: 'IndexLabel' = None, inplace: 'bool' = False, ignore_index: 'bool' = False) -> 'DataFrame | None' method of pandas.core.frame.DataFrame instance
    Remove missing values.
    
    See the :ref:`User Guide <missing_data>` for more on which values are
    considered missing, and how to work with missing data.
    
    Parameters
    ----------
    axis : {0 or 'index', 1 or 'columns'}, default 0
        Determine if rows or columns which contain missing values are
        removed.
    
        * 0, or 'index' : Drop rows which contain missing values.
        * 1, or 'columns' : Drop columns which contain missing value.
    
        Pass tuple or list to drop on multiple axes.
        Only a single axis is allowed.
    
    how : {'any', 'all'}, default 'any'
        Determine if row or column is removed from DataFrame

In [ ]:
"""
EXERCICE - La méthode dropna
La méthode dropna permet d'éliminer les valeurs manquantes (NaN). Lisez d'abord sa documentation.
1. À quoi correspondent les arguments "axis" et "how" ?
2. Éliminez toutes les lignes contenant uniquement des valeurs manquantes.
3. Éliminez toutes les lignes contenant au moins une valeur manquante. Combien y a-t-il de lignes de différence ?
4. Éliminez toutes les colonnes contenant au moins une valeur manquante.
"""

"""
RÉPONSES - La méthode dropna
"""
#1) Si axis=0, on supprime la ligne qui contient une valeur manquante.
#   Si axis=1, on supprime la colonne qui contient une valeur manquante.
#   Si how='any', on supprime la colonne  ou la ligne si elle contient au moins une valeur manquante. 
#   Si how='all', on supprime la colonne  ou la ligne si toutes ses valeurs sont des valeurs manquantes.

# 2. Éliminons toutes les lignes contenant uniquement des valeurs manquantes.

df2=df2.dropna(axis=0, how='all')

#3. Éliminons toutes les lignes contenant au moins une valeur manquante. Combien y a-t-il de lignes de différence ?
df2=df2.dropna(axis=0, how='any')

print(f"La différence de ligne après ces deux suppressions est: \n ", df.shape[1]-df2.shape[1])

#4. Éliminons toutes les colonnes contenant au moins une valeur manquante.

df2=df2.dropna(axis=1, how='any')


La différence de ligne après ces deux suppressions est: 
  1


In [ ]:
help(df.fillna)

Help on method fillna in module pandas.core.frame:

fillna(value: 'Hashable | Mapping | Series | DataFrame' = None, *, method: 'FillnaOptions | None' = None, axis: 'Axis | None' = None, inplace: 'bool' = False, limit: 'int | None' = None, downcast: 'dict | None' = None) -> 'DataFrame | None' method of pandas.core.frame.DataFrame instance
    Fill NA/NaN values using the specified method.
    
    Parameters
    ----------
    value : scalar, dict, Series, or DataFrame
        Value to use to fill holes (e.g. 0), alternately a
        dict/Series/DataFrame of values specifying which value to use for
        each index (for a Series) or column (for a DataFrame).  Values not
        in the dict/Series/DataFrame will not be filled. This value cannot
        be a list.
    method : {'backfill', 'bfill', 'ffill', None}, default None
        Method to use for filling holes in reindexed Series:
    
        * ffill: propagate last valid observation forward to next valid.
        * backfill / b

In [ ]:
"""
EXERCICE - La méthode fillna
La méthode dropna permet d'imputer les valeurs manquantes (NaN). Lisez d'abord sa documentation.
1. Quelles sont les différentes stratégies de remplissage des valeurs manquantes ?
2. Imputez les valeurs manquantes de la colonne age du DF donné en exemple par :
    - 0
    - la dernière valeur précédente/suivante valide
    - la moyenne
    - la valeur la plus courante (mode)
3. Quel est le meilleur choix dans ce cas ? Et pour le cas d'une variable temporelle, par exemple la température 'EGT_SEL' de notre jeu de données ?
"""
exemple = pd.DataFrame({'nom': ['Alice', 'Bob', 'Charlie', 'David'], 'age': [24, np.nan, 99, 24]})

#1. les différentes stratégies de remplissage des valeurs manquantes sont: 
#   % remplissage par la valeurs précédente (par ligne ou par colonne)
#   % remplissage par la valeurs suivante (par ligne ou par colonne)

#2. Imputons les valeurs manquantes de la colonne age du DF donné en exemple par :
    #- 0
    
exemple1 = exemple
exemple1=exemple1.fillna(value=0)
print(f"Remplacement avec '0': \n", exemple1)
  
    #- la dernière valeur précédente/suivante valide
exemple2 = exemple
exemple2=exemple2.fillna(method='backfill') 
print(f"Remplacement avec la dernière valeur suivante valide: \n", exemple2)
exemple3 = exemple
exemple3=exemple3.fillna(method='ffill') 
print(f"Remplacement avec la dernière valeur précédente valide: \n", exemple3)   

    # - la moyenne
exemple4=exemple.fillna(value=exemple['age'].mean())
print(f"Remplacement avec la moyenne : \n", exemple4)
  
    #- la valeur la plus courante (mode)
exemple5=exemple.fillna(value=float(exemple['age'].mode().iloc[0]))
print(f"Remplacement avec la moyenne : \n", exemple5)

# le meilleur choix dans ce cas est le remplacement par la moyenne

#Et pour le cas d'une variable temporelle, par exemple la température 'EGT_SEL' de notre jeu de données, on remplacera par la dernière valeur précédente valide.

Remplacement avec '0': 
        nom   age
0    Alice  24.0
1      Bob   0.0
2  Charlie  99.0
3    David  24.0
Remplacement avec la dernière valeur suivante valide: 
        nom   age
0    Alice  24.0
1      Bob  99.0
2  Charlie  99.0
3    David  24.0
Remplacement avec la dernière valeur précédente valide: 
        nom   age
0    Alice  24.0
1      Bob  24.0
2  Charlie  99.0
3    David  24.0
Remplacement avec la moyenne : 
        nom   age
0    Alice  24.0
1      Bob  49.0
2  Charlie  99.0
3    David  24.0
Remplacement avec la moyenne : 
        nom   age
0    Alice  24.0
1      Bob  24.0
2  Charlie  99.0
3    David  24.0


t
2011-09-15 14:26:02.875    335.9840
2011-09-15 14:26:03.125    239.9860
2011-09-15 14:26:03.375    167.9910
2011-09-15 14:26:03.625    119.9940
2011-09-15 14:26:03.875     72.0011
                             ...   
2011-09-15 16:01:19.875    230.8250
2011-09-15 16:01:20.125    137.4770
2011-09-15 16:01:20.375    113.4530
2011-09-15 16:01:20.625    158.8080
2011-09-15 16:01:20.875    204.1800
Name: EGT_SEL, Length: 22825, dtype: float64

## Git

Créer un compte GitHub : https://github.com/

Créer un nouveau repository :
1. Sélectionner "Nouveau dépôt" dans le menu déroulant avec le signe +. 
2. Saisisser un nom pour votre dépôt (par exemple, "TDs statistiques descriptives") 
3. Cliquer sur "Créer un dépôt". Ne vous souciez pas des autres options.



Télécharger Git Bash : https://git-scm.com/downloads

Sur Git Bash, positionnez-vous dans le répertoire où votre projet figure.

Ensuite

> git init

> git remote add origin https://github.com/********

> git add .

> git commit -m "Ajout du TD1"

> git push origin master

C'est tout ! Vous avez créé votre repo GitHub ! 